main/src/library/LazyList.flix

```diff
@@ -0,0 +1,242 @@
+/*
+ * Copyright 2020 Esben Bjerre
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace LazyList {
+
+    ///
+    /// An immutable lazy list.
+    ///
+    /// A lazy list is either the empty list represented by `Empty`, or
+    /// an element `x` followed by a lazy list `xs`.
+    /// Note that a lazy list is lazy on its tail but strict on its head.
+    /// The empty list is always strict.
+    ///
+    pub enum LazyList[t] {
+        case Empty,
+        case LazyCons(t, Lazy[LazyList[t]])
+    }
+
+    ///
+    /// Returns the empty LazyList.
+    ///
+    pub def empty(): LazyList[a] = Empty
+
+    ///
+    /// Returns an infinite LazyList of `x`.
+    ///
+    pub def repeat(x: a): LazyList[a] = LazyCons(x, lazy repeat(x))
+
+    ///
+    /// Returns an infinite sequence of integers starting from and including `n`.
+    ///
+    pub def from(n: Int32): LazyList[Int32] = LazyCons(n, lazy from(n + 1))
```

Comment:
Yes that is the idea. Your example would run forever since from is infinite, but something like `LazyList.length(LazyList.take(100000, LazyList.from(0)))` returns 100000 without overflowing the stack.